### PR TITLE
Sparkle effect

### LINKED ISSUE #19: [Effect] Sparkling Particles
Become larger and brighter

### PR DESCRIPTION
**Pull request by suluke**
This will fix #19 

I will implement UI later

### CODE REVIEW

File: js/command-builder.js [hunk 1]
```diff
@@ -153,13 +145,15 @@ export default class CommandBuilder {
           return res();
         }
         const effectUniforms = new Uniforms(globalId);
+        const effectAttributes = new Attributes(globalId);
         const effectClass = effectConfig.getEffectClass();
         vert.mainBody += `if (${effectConfig.timeBegin} <= globalTime && globalTime <= ${effectConfig.timeEnd}) {`;
-        effectClass.registerAsync(effectConfig, this.props, effectUniforms, vert)
+        effectClass.registerAsync(effectConfig, this.props, effectUniforms, vert, frag, effectAttributes)
```
Comment (suluke) [inline js/command-builder.js:151]: Passing the fragment shader into `register` was of course not necessary. But I believed it was in the beginning, so now it's there and I don't think it's a bad thing. But I can of course revert it

File: js/effects/effect.js [hunk 2]
```diff
@@ -2,12 +2,12 @@
  * Interface for effects
  */
 export default class Effect {
-  static register(/* instance, props, uniforms, vertexShader */) {
+  static register(/* instance, props, uniforms, vertexShader, fragmentShader */) {
```
Comment (suluke) [inline js/effects/effect.js]: I forgot to add the `attributes` param here

File: js/renderer/state.js [hunk 3]
```diff
@@ -101,6 +101,7 @@ export default class RendererState {
     // Properties
     this.particleData = -1;
     this.particleDataStore = [[null, null]];
+    this.buffers = [];
```
Comment (suluke) [inline js/renderer/state.js:104]: Let's see if you are ok with `RendererState` doing buffer management :grin:

File: js/effects/sparkle.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+      if (instance.timeBegin <= time && time <= instance.timeEnd) {
+        for (let i = 0; i < progresses.length; i++) {
+          if (progresses[i] != 0 || shouldStart()) {
+            progresses[i] += props.clock.getDelta() / sparkleTime;
+            if (progresses[i] > 1) {
+              progresses[i] = 0;
+            }
+          }
+        }
+        buffer({ data: progresses });
```
Comment (suluke) [inline js/effects/sparkle.js]: Looping over the data and changing it for each frame seems possibly a bit inefficient - not sure. If you have a better idea, I'm all open for it

**Comment by luckyxxl** [comment]
As you might have imagined, I'm objected against doing per-frame per-particle calculations on the cpu. I suggest an alternative: At register, fill the attribute with random data defining a random "phase-offset" and "period" based on the probabilities you derived and let the vertex shader use this data to calculate the progress. While this is not strictly random anymore I think this is fine since the random "period" should introduce enough cross-particle randomness so it will be hard to recognize that each particle itself is not random.

Concerning buffer management in RendererState: I'm totally ok with that, but how does it work with effects possibly changing the particle count (FlickrImage and possible future ChangeParticleCount)?

Apart from that the pull request looks fine, no objections here ;)

Oh, and I really like the effect ;)

**Comment by luckyxxl** [comment]
What do you think about this commit? With this I think we can merge.

**Comment by suluke** [comment]
@luckyxxl I'm not yet happy to merge this. I have the impression your commit was hacked together very quickly. Both formatting as well as the logic itself seem to have flaws. Also, I have trouble understanding the overall concept, so I think some comments would greatly improve the quality of the code.

**Comment by luckyxxl** [comment]
@suluke Ok, I fixed several things you mentioned in the review. The inverted x was a misunderstanding by me. I hope I'm using the duration parameter as originally intended by you?
After writing the explanation in the code, I see now that maybe the code could be simplified by using a real sawtooth function, but I think that this "offsetted-sawtooth" is easier to understand.
One thing left to do is the instance.repetitions parameter. I can imagine that simply dividing duration by this parameter yields the expected results (although not strictly correct).
Yes, the code might look a bit hacked since I had several problems during the implementation 0:)

**Pull request merged**